docs/design/coreclr/botr/vectors-and-intrinsics.md

```diff
@@ -0,0 +1,189 @@
+Vectors and Hardware Intrinsics Support
+===
+---
+
+# Introduction
+The CoreCLR runtime has support for several varieties of hardware intrinsics, and various ways to compile code which uses them. This support varies by target processor, and the code produced depends on how the jit compiler is invoked. This document describes the various behaviors of intrinsics in the runtime, and concludes with implications  for developers working on the runtime and libraries portions of the runtime. 
+
+# Intrinsics apis
+Most hardware intrinsics support is tied to the use of various Vector apis. There are 4 major api surfaces that are supported by the runtime
+
+- The fixed length float vectors. `Vector2`, `Vector3`, and `Vector4`. These vector types represent a struct of floats of various lengths. For type layout, ABI and, interop purposes they are represented in exactly the same way as a structure with an appropriate number of floats in it. Operations on these vector types are supported on all architectures and platforms, although some architectures may optimize various operations.
+- The variable length `Vector<T>`. This represents vector data of runtime-determined length. In any given process the length of a `Vector<T>` is the same in all methods, but this length may differ between various machines or environment variable settings read at startup of the process. The `T` type variable may be any of the primitive types, and allows use of integer or double data within a vector. The length and alignment of `Vector<T>` is unknown to the developer, and `Vector<T>` may not exist in any interop signature. Operations on these vector types are supported on all architectures and platforms, although some architectures may optimize various operations.
```

Comment:
> The `T` type variable may be any of the primitive types

Maybe this could be reworded? It isn't allowed to be `char` or `bool` for example. Only the `8` basic integer types (signed and unsigned `int8`, `int16`, `int32`, and `int64`) and the `2` basic floating-point types (`float32` and `float64`)

Reply:
> The length and alignment of `Vector<T>` is unknown to the developer, and `Vector<T>` may not exist in any interop signature

This isn't quite true, it is known at `Runtime` by checking `Vector<T>.Count` which specifies it is a struct of exactly `Count` `T` fields.